File: spec/bullets_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe 'Bullets.vim' do
  describe 'inserting new bullets' do
    context 'on return key when cursor is not at EOL' do
      it 'splits the line and does not add a bullet' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          - this is the first bullet
        TEXT

        vim.edit filename
        vim.type 'G$i'
        vim.feedkeys '\<cr>'
        vim.type 'second bullet'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          - this is the first bulle
          second bullett\n
        TEXT
      end
    end

    context 'on return key when cursor is at EOL' do
      it 'adds a new bullet if the previous line had a known bullet type' do
        test_bullet_inserted('do that', <<-INIT, <<-EXPECTED)
          # Hello there
          - do this
        INIT
          # Hello there
          - do this
          - do that
        EXPECTED
      end

      it 'adds a new latex bullet' do
        test_bullet_inserted('Second item', <<-INIT, <<-EXPECTED)
        \\documentclass{article}
          \\begin{document}

          \\begin{itemize}
            \\item First item
        INIT
        \\documentclass{article}
          \\begin{document}

          \\begin{itemize}
            \\item First item
            \\item Second item
        EXPECTED
      end

      it 'adds a pandoc bullet if the prev line had one' do
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          Hello there
          #. this is the first bullet
        INIT
          Hello there
          #. this is the first bullet
          #. second bullet
        EXPECTED
      end

      it 'adds an Org mode bullet if the prev line had one' do
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          Hello there
          **** this is the first bullet
        INIT
          Hello there
          **** this is the first bullet
          **** second bullet
        EXPECTED
      end

      it 'adds a new numeric bullet if the previous line had numeric bullet' do
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          # Hello there
          1) this is the first bullet
        INIT
          # Hello there
          1) this is the first bullet
          2) second bullet
        EXPECTED
      end

      it 'adds a new numeric bullet with right padding' do
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          # Hello there
          1.  this is the first bullet
        INIT
          # Hello there
          1.  this is the first bullet
          2.  second bullet
        EXPECTED
      end

      it 'maintains total bullet width from 9. to 10. with reduced padding' do
        vim.command('let g:bullets_renumber_on_change=0')
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          # Hello there
          9.  this is the first bullet
        INIT
          # Hello there
          9.  this is the first bullet
          10. second bullet
        EXPECTED
      end

      it 'adds a new - bullet with right padding' do
        test_bullet_inserted('second bullet', <<-INIT, <<-EXPECTED)
          # Hello there
          -   this is the first bullet
        INIT
          # Hello there
          -   this is the first bullet
          -   second bullet
        EXPECTED
      end

      it 'does not insert a new numeric bullet for decimal numbers' do
        test_bullet_inserted('second line', <<-INIT, <<-EXPECTED)
          # Hello there
          3.14159 is an approximation of pi.
        INIT
          # Hello there
          3.14159 is an approximation of pi.
          second line
        EXPECTED
      end

      it 'adds a new roman numeral bullet' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          I. this is the first bullet
        TEXT

        vim.command 'let g:bullets_pad_right = 0'
        vim.edit filename
        vim.type 'GA'
        vim.feedkeys '\<cr>'
        vim.type 'second bullet'
        vim.feedkeys '\<cr>'
        vim.type 'third bullet'
        vim.feedkeys '\<cr>'
        vim.type 'fourth bullet'
        vim.feedkeys '\<cr>'
        vim.type 'fifth bullet'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          I. this is the first bullet
          II. second bullet
          III. third bullet
          IV. fourth bullet
          V. fifth bullet\n
        TEXT
      end

      it 'adds a new lowercase roman numeral bullet' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          i. this is the first bullet
        TEXT

        vim.command 'let g:bullets_pad_right = 0'
        vim.edit filename
        vim.type 'GA'
        vim.feedkeys '\<cr>'
        vim.type 'second bullet'
        vim.feedkeys '\<cr>'
        vim.type 'third bullet'
        vim.feedkeys '\<cr>'
        vim.type 'fourth bullet'
        vim.feedkeys '\<cr>'
        vim.type 'fifth bullet'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          i. this is the first bullet
          ii. second bullet
          iii. third bullet
          iv. fourth bullet
          v. fifth bullet\n
        TEXT
      end

      it 'does not confuse with the "ignorecase" option' do
        vim.command 'set ignorecase'
        test_bullet_inserted('second line', <<-INIT, <<-EXPECTED)
          # Hello there
          Vi. this is the first line
        INIT
          # Hello there
          Vi. this is the first line
          second line
        EXPECTED
      end

      it 'does not insert a new roman bullets without following spaces' do
        test_bullet_inserted('second line', <<-INIT, <<-EXPECTED)
          # Hello there
          m.example.com is a site.
        INIT
          # Hello there
          m.example.com is a site.
          second line
        EXPECTED
      end

      it 'does not insert a new roman bullets for invalid roman numbers' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          LID. the first line
        TEXT

        vim.edit filename
        vim.type 'GA'
        vim.feedkeys '\<cr>'
        vim.type 'second line'
        vim.feedkeys '\<cr>'
        vim.type 'vim. third line'
        vim.feedkeys '\<cr>'
        vim.type 'fourth line'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          LID. the first line
          second line
          vim. third line
          fourth line\n
        TEXT
      end

      it 'deletes the last bullet if it is empty' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          - this is the first bullet
        TEXT

        vim.edit filename
        vim.type 'GA'
        vim.feedkeys '\<cr>'
        vim.feedkeys '\<cr>'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents.strip).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          - this is the first bullet
        TEXT
      end

      it 'does not delete the last bullet when configured not to' do
        filename = "#{SecureRandom.hex(6)}.txt"
        write_file(filename, <<-TEXT)
          # Hello there
          - this is the first bullet
        TEXT

        vim.command 'let g:bullets_delete_last_bullet_if_empty = 0'
        vim.edit filename
        vim.type 'GA'
        vim.feedkeys '\<cr>'
        vim.feedkeys '\<cr>'
        vim.write

        file_contents = IO.read(filename)

        expect(file_contents.strip).to eq normalize_string_indent(<<-TEXT)
          # Hello there
          - this is the first bullet
          -
        TEXT
      end
    end
  end
end

File: test/bullets_spec.lua
local helpers = require("test.helpers")

describe("Bullets.vim", function()
  describe("inserting new bullets", function()
    before_each(function()
      helpers.reset_config()
      vim.o.ignorecase = false
    end)

    describe("on return key when cursor is not at EOL", function()
      it("splits the line and does not add a bullet", function()
        -- G$i places cursor on last char 't' in "- this is the first bullet"
        -- i inserts before 't', CR is deferred (not at EOL), splits the line
        -- then "second bullet" is typed before 't': "second bullett"
        -- Temporarily disable formatoptions 'r' to avoid comment-leader insertion
        local saved_fo = vim.o.formatoptions
        vim.cmd("set formatoptions-=r")
        helpers.new_buffer({
          "# Hello there",
          "- this is the first bullet",
        })
        -- Position cursor on 't' (last char, 0-indexed col 25), enter insert before it
        -- CR is deferred by plugin (not at EOL); 'x' flag drains the deferred CR → split
        -- We end up in normal mode on new line with 't'
        helpers.feedkeys("G$i<CR>")
        -- Now in normal mode at col 0 of "t"; insert before 't' and type
        helpers.feedkeys("isecond bullet<Esc>")
        vim.o.formatoptions = saved_fo
        assert.are.same({
          "# Hello there",
          "- this is the first bulle",
          "second bullett",
        }, helpers.get_lines())
      end)
    end)

    describe("on return key when cursor is at EOL", function()
      it("adds a new bullet if the previous line had a known bullet type", function()
        helpers.test_bullet_inserted(
          "do that",
          { "# Hello there", "- do this" },
          { "# Hello there", "- do this", "- do that" }
        )
      end)

      it("adds a new latex bullet", function()
        helpers.test_bullet_inserted("Second item", {
          "\\documentclass{article}",
          "  \\begin{document}",
          "",
          "  \\begin{itemize}",
          "    \\item First item",
        }, {
          "\\documentclass{article}",
          "  \\begin{document}",
          "",
          "  \\begin{itemize}",
          "    \\item First item",
          "    \\item Second item",
        })
      end)

      it("adds a pandoc bullet if the prev line had one", function()
        helpers.test_bullet_inserted(
          "second bullet",
          { "Hello there", "#. this is the first bullet" },
          { "Hello there", "#. this is the first bullet", "#. second bullet" }
        )
      end)

      it("adds an Org mode bullet if the prev line had one", function()
        helpers.test_bullet_inserted(
          "second bullet",
          { "Hello there", "**** this is the first bullet" },
          { "Hello there", "**** this is the first bullet", "**** second bullet" }
        )
      end)

      it("adds a new numeric bullet if the previous line had numeric bullet", function()
        helpers.test_bullet_inserted(
          "second bullet",
          { "# Hello there", "1) this is the first bullet" },
          { "# Hello there", "1) this is the first bullet", "2) second bullet" }
        )
      end)

      it("adds a new numeric bullet with right padding", function()
        helpers.test_bullet_inserted(
          "second bullet",
          { "# Hello there", "1.  this is the first bullet" },
          { "# Hello there", "1.  this is the first bullet", "2.  second bullet" }
        )
      end)

      it("maintains total bullet width from 9. to 10. with reduced padding", function()
        vim.g.bullets_renumber_on_change = 0
        helpers.test_bullet_inserted(
          "second bullet",
          { "# Hello there", "9.  this is the first bullet" },
          { "# Hello there", "9.  this is the first bullet", "10. second bullet" }
        )
      end)

      it("adds a new - bullet with right padding", function()
        helpers.test_bullet_inserted(
          "second bullet",
          { "# Hello there", "-   this is the first bullet" },
          { "# Hello there", "-   this is the first bullet", "-   second bullet" }
        )
      end)

      it("does not insert a new numeric bullet for decimal numbers", function()
        -- "3.14159 is an approximation of pi." is not a bullet line
        -- CR on non-bullet line is deferred, use two-call pattern
        helpers.new_buffer({
          "# Hello there",
          "3.14159 is an approximation of pi.",
        })
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("isecond line<Esc>")
        assert.are.same({
          "# Hello there",
          "3.14159 is an approximation of pi.",
          "second line",
        }, helpers.get_lines())
      end)

      it("adds a new roman numeral bullet", function()
        vim.g.bullets_pad_right = 0
        helpers.new_buffer({
          "# Hello there",
          "I. this is the first bullet",
        })
        helpers.feedkeys("A<CR>second bullet<CR>third bullet<CR>fourth bullet<CR>fifth bullet<Esc>")
        assert.are.same({
          "# Hello there",
          "I. this is the first bullet",
          "II. second bullet",
          "III. third bullet",
          "IV. fourth bullet",
          "V. fifth bullet",
        }, helpers.get_lines())
      end)

      it("adds a new lowercase roman numeral bullet", function()
        vim.g.bullets_pad_right = 0
        helpers.new_buffer({
          "# Hello there",
          "i. this is the first bullet",
        })
        helpers.feedkeys("A<CR>second bullet<CR>third bullet<CR>fourth bullet<CR>fifth bullet<Esc>")
        assert.are.same({
          "# Hello there",
          "i. this is the first bullet",
          "ii. second bullet",
          "iii. third bullet",
          "iv. fourth bullet",
          "v. fifth bullet",
        }, helpers.get_lines())
      end)

      it("does not confuse with the 'ignorecase' option", function()
        vim.cmd("set ignorecase")
        -- "Vi." is mixed case / not a valid roman numeral bullet → non-bullet CR
        helpers.new_buffer({
          "# Hello there",
          "Vi. this is the first line",
        })
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("isecond line<Esc>")
        assert.are.same({
          "# Hello there",
          "Vi. this is the first line",
          "second line",
        }, helpers.get_lines())
      end)

      it("does not insert a new roman bullets without following spaces", function()
        -- "m.example.com is a site." has no space after the dot → not a bullet
        helpers.new_buffer({
          "# Hello there",
          "m.example.com is a site.",
        })
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("isecond line<Esc>")
        assert.are.same({
          "# Hello there",
          "m.example.com is a site.",
          "second line",
        }, helpers.get_lines())
      end)

      it("does not insert a new roman bullets for invalid roman numbers", function()
        -- "LID." is not a valid roman numeral, so no bullet continuation
        -- However lines typed after non-bullet lines also get no bullet
        helpers.new_buffer({
          "# Hello there",
          "LID. the first line",
        })
        -- First CR after "LID. the first line" (non-bullet) → deferred
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("isecond line<Esc>")
        -- Now on "second line" (non-bullet) → deferred CR
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("ivim. third line<Esc>")
        -- Now on "vim. third line" (non-bullet) → deferred CR
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("ifourth line<Esc>")
        assert.are.same({
          "# Hello there",
          "LID. the first line",
          "second line",
          "vim. third line",
          "fourth line",
        }, helpers.get_lines())
      end)

      it("deletes the last bullet if it is empty", function()
        helpers.new_buffer({
          "# Hello there",
          "- this is the first bullet",
        })
        -- First CR creates "- " empty bullet, second CR on empty bullet deletes it
        helpers.feedkeys("A<CR><CR>")
        local lines = helpers.get_lines()
        -- Strip trailing empty lines before comparison.
        while #lines > 0 and lines[#lines] == "" do
          table.remove(lines)
        end
        assert.are.same({
          "# Hello there",
          "- this is the first bullet",
        }, lines)
      end)

      it("promote the last bullet when configured to", function()
        vim.g.bullets_delete_last_bullet_if_empty = 2
        helpers.new_buffer({
          "# Hello there",
          "- this is the first bullet",
          "  - this is the second bullet",
        })
        -- First CR creates new child bullet "  - ", second CR promotes (dedents)
        helpers.feedkeys("A<CR><CR>")
        local lines = helpers.get_lines()
        -- strip trailing empty lines
        while #lines > 0 and lines[#lines] == "" do
          table.remove(lines)
        end
        -- The promoted bullet has a trailing space ("- ") matching plugin output
        assert.are.same({
          "# Hello there",
          "- this is the first bullet",
          "  - this is the second bullet",
          "- ",
        }, lines)
      end)

      it("does not delete the last bullet when configured not to", function()
        vim.g.bullets_delete_last_bullet_if_empty = 0
        helpers.new_buffer({
          "# Hello there",
          "- this is the first bullet",
        })
        -- First CR creates "- " empty bullet, second CR on empty bullet
        -- with config=0 does NOT delete it; a plain CR fires leaving "- " intact
        helpers.feedkeys("A<CR><CR>")
        local lines = helpers.get_lines()
        -- strip trailing empty lines
        while #lines > 0 and lines[#lines] == "" do
          table.remove(lines)
        end
        -- The non-deleted bullet retains trailing space ("- ") matching plugin output
        assert.are.same({
          "# Hello there",
          "- this is the first bullet",
          "- ",
        }, lines)
      end)

      it("toggles roman numeral bullets with g:bullets_enable_roman_list", function()
        -- Disable alpha lists to isolate test to roman numerals
        vim.g.bullets_max_alpha_characters = 0
        vim.g.bullets_enable_roman_list = 1
        helpers.new_buffer({
          "# Hello there",
          "i. this is the first bullet",
        })
        -- Type second and third bullets (roman numeral bullets)
        helpers.feedkeys("A<CR>second bullet<CR>third bullet<Esc>")
        -- Disable roman list mid-test
        vim.g.bullets_enable_roman_list = 0
        -- Type fourth and fifth (no roman numeral prefix now)
        -- We're in normal mode, need to append and continue
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("ifourth bullet<Esc>")
        helpers.feedkeys("A<CR>")
        helpers.feedkeys("ififth bullet<Esc>")
        assert.are.same({
          "# Hello there",
          "i. this is the first bullet",
          "ii. second bullet",
          "iii. third bullet",
          "fourth bullet",
          "fifth bullet",
        }, helpers.get_lines())
      end)
    end)
  end)
end)
